app/code/community/Bolt/Boltpay/OrderCreationException.php

```diff
@@ -0,0 +1,214 @@
+    const E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_EXPIRED    = '{"reason": "This coupon has expired", "discount_code": "%s"}';
+    const E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_GENERIC    = '{"reason": "%s", "discount_code": "%s"}';
+
+    /**
+     * Invalid discount code used
+     */
+    const E_BOLT_DISCOUNT_DOES_NOT_EXIST        = 2001007;
+    const E_BOLT_DISCOUNT_DOES_NOT_EXIST_TMPL   = '{"discount_code": "%s"}';
+
+    /**
+     * Change in shipping price
+     */
+    const E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED           = 2001008;
+    const E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED_TMPL      = '{"reason": "Shipping total has changed", "old_value": "%d", "new_value": "%d"}';
+
+    /**
+     * @var int http response code that is to be returned
+     */
+    protected $httpCode;
+
+    /**
+     * @var string The json to be returned to Bolt associated with this exception
+     */
+    protected $json;
+
+    /**
+     * @var array All possible error codes supported by Bolt
+     */
+    protected static $validCodes = array(
+        self::E_BOLT_GENERAL_ERROR,
+        self::E_BOLT_ORDER_ALREADY_EXISTS,
+        self::E_BOLT_CART_HAS_EXPIRED,
+        self::E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED,
+        self::E_BOLT_OUT_OF_INVENTORY,
+        self::E_BOLT_DISCOUNT_CANNOT_APPLY,
+        self::E_BOLT_DISCOUNT_DOES_NOT_EXIST,
+        self::E_BOLT_SHIPPING_PRICE_HAS_BEEN_UPDATED
+    );
+
+    /**
+     * Bolt_Boltpay_InvalidTransitionException constructor.
+     *
+     * @param int    $code          The Bolt defined error code
+     * @param string $dataTemplate  specific Bolt error sub-category
+     * @param array  $dataValues    An array of values to be added to the data template
+     * @param string $message       The exception message to throw.
+     * @param Throwable|null $previous  [optional] The previously throwable used for exception chaining.
+     */
+    public function __construct($code = self::E_BOLT_GENERAL_ERROR, $dataTemplate = self::E_BOLT_GENERAL_ERROR_TMPL_GENERIC, array $dataValues = array(), $message = null, Throwable $previous = null)
+    {
+        // If code is invalid, we will use the generic message
+        if (!in_array($code, self::$validCodes)) {
+            $code = self::E_BOLT_GENERAL_ERROR;
+            $dataTemplate = self::E_BOLT_GENERAL_ERROR_TMPL_GENERIC;
+            $dataValues = array(addcslashes($message, '"\\'));
+        }
+
+        $this->setHttpCode($code, $dataTemplate);
+        $this->setJson($code, $dataTemplate, $dataValues);
+
+        if (empty($message)) {
+            $message = $this->getJson();
+        }
+
+        parent::__construct($message, $code, $previous);
+    }
+
+    /**
+     * Sets the httpCode based on the error
+     *
+     * @param int    $code         Bolt error code
+     * @param string $dataTemplate specific Bolt error sub-category
+     *
+     * @return int The HTTP code that was set
+     */
+    private function setHttpCode( $code, $dataTemplate ) {
```

Comment:
I think readability would increase if you split logic to get code. eg., you can define `getCode` method that just returns code

```
private function getHttpCode( $code, $dataTemplate ) {
switch ($code) {
            case self::E_BOLT_GENERAL_ERROR:
                if ($dataTemplate === self::E_BOLT_GENERAL_ERROR_TMPL_HMAC) {
                    return 401;
                }
            case self::E_BOLT_ORDER_ALREADY_EXISTS:
                return 409;
            }
}
```

and then instead of `$this->setHttpCode` you can do `$this->httpCode = $this->getHttpCode`

Reply:
or instead of `getHttpCode` maybe you can have map from error code to http status code

Reply:
There are many good approaches to doing many things.  We must makes sure that are assuring that a good, reasonable approach is used, which is not always our own. 

This is an stylistic request that I agree with.  Having the class data members visibly set in the constructor makes for a faster easier read.

A map is not a good approach.  If the business logic expands, the map will prove to be limiting.  Currently, you can see this for case `self::E_BOLT_GENERAL_ERROR` and `default`.  If the code maps to `self::E_BOLT_GENERAL_ERROR`, then only in the case that the template is for `self::E_BOLT_GENERAL_ERROR_TMPL_HMAC` should 401 be returned, otherwise the default 422.  There is construct for default values in a associative array in PHP, so the switch/case construct is optimal for this use case.

Task created here:
https://app.asana.com/0/1118494470563243/1121207201774762